darshan-util/pydarshan/darshan/experimental/plots/heatmap_handling.py

```diff
@@ -0,0 +1,442 @@
+        # read in the module data, update the name records
+        report.mod_read_all_dxt_records(module_key, dtype="pandas")
+        # retrieve the list of records in pd.DataFrame() form
+        dict_list = report.records[module_key].to_df()
+        # retrieve the list of read/write dataframes from the list of records
+        read_df, write_df = get_rd_wr_dfs(dict_list=dict_list)
+        # create empty dictionary for each module
+        flat_data_dict[module_key] = {}
+        if "read" in ops:
+            # add the concatenated dataframe to the flat dictionary
+            flat_data_dict[module_key]["read"] = read_df
+        if "write" in ops:
+            # add the concatenated dataframe to the flat dictionary
+            flat_data_dict[module_key]["write"] = write_df
+
+    return flat_data_dict
+
+
+def get_aggregate_data(
+    report: Any,
+    mods: Sequence[str] = ["DXT_POSIX"],
+    ops: Sequence[str] = ["read", "write"],
+) -> pd.DataFrame:
+    """
+    Aggregates the data based on which
+    modules and operations are selected.
+
+    Parameters
+    ----------
+
+    report: a ``darshan.DarshanReport``.
+
+    mods: a list of keys designating which Darshan modules to use for
+    data aggregation. Default is ``["DXT_POSIX"]``.
+
+    ops: a list of keys designating which Darshan operations to use for
+    data aggregation. Default is ``["read", "write"]``.
+
+    Returns
+    -------
+
+    agg_df: a ``pd.DataFrame`` containing the aggregated data determined
+    by the input modules and operations.
+
+    Raises
+    ------
+
+    ValueError: raised if the selected modules/operations
+    don't contain any data.
+
+    Notes
+    -----
+    Since read and write events are considered unique events, if both are
+    selected their dataframes are simply concatenated.
+
+    Examples
+    --------
+    `agg_df` generated from `tests/input/sample-dxt-simple.darshan`:
+
+            length  start_time  end_time  rank
+        0      40    0.103379  0.103388     0
+        1    4000    0.104217  0.104231     0
+
+    """
+    # collect the concatenated dataframe data from the darshan report
+    df_dict = get_single_df_dict(report=report, mods=mods, ops=ops)
+    # TODO: generalize for all DXT modules, for now manually set `DXT_POSIX`
+    module_key = "DXT_POSIX"
+    # iterate over each dataframe based on which operations are selected
+    df_list = []
+    for op_key, op_df in df_dict[module_key].items():
+        # if the dataframe has data, append it to the list
+        if op_df.size:
+            df_list.append(op_df)
+
+    if df_list:
+        # if there are dataframes in the list, concatenate them into 1 dataframe
+        agg_df = pd.concat(df_list, ignore_index=True)
+    else:
+        raise ValueError(f"No data available for selected module(s) and operation(s).")
+
+    return agg_df
+
+
+def calc_prop_data_sum(
+    tmin: float,
+    tmax: float,
+    total_elapsed: np.ndarray,
+    total_data: np.ndarray,
```

Comment:
If we're going to start typing with NumPy we'll probably want to use: https://numpy.org/devdocs/reference/typing.html

It is pretty bleeding edge stuff, but it means we can type the data inside the array as well--an example is here:
https://github.com/scipy/scipy/pull/14281

So you get stuff like: `points: npt.NDArray[np.float64]`

Reply:
Probably needs a guard like this as well because it is so new it will only be executable in a separate check that has newer NumPy that we can't expect folks to have in general:

```python
if TYPE_CHECKING:
    import numpy.typing as npt
```

Similar to this I suppose: https://github.com/scipy/scipy/pull/14174/files

Reply:
I was able to update these using the `numpy.typing` module and add in the `if TYPE_CHECKING` block. This should be resolved now.